File: backend/src/main/java/com/safebite/backend/service/AlertaService.java
package com.safebite.backend.service;

import com.safebite.backend.dto.request.AlertaRequest;
import com.safebite.backend.dto.request.RevisarAlertaRequest;
import com.safebite.backend.dto.response.AlertaResponse;
import com.safebite.backend.exception.BadRequestException;
import com.safebite.backend.exception.ResourceNotFoundException;
import com.safebite.backend.model.Alerta;
import com.safebite.backend.model.EstadoAlerta;
import com.safebite.backend.model.Producto;
import com.safebite.backend.model.Restaurante;
import com.safebite.backend.model.Usuario;
import com.safebite.backend.repository.AlertaRepository;
import com.safebite.backend.repository.ProductoRepository;
import com.safebite.backend.repository.RestauranteRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class AlertaService {

    private final AlertaRepository alertaRepository;
    private final RestauranteRepository restauranteRepository;
    private final ProductoRepository productoRepository;

    @Transactional
    public AlertaResponse crear(AlertaRequest request, Usuario usuario) {
        Restaurante restaurante = null;
        Producto producto = null;

        if (request.getRestauranteId() != null) {
            restaurante = restauranteRepository.findById(request.getRestauranteId())
                    .orElseThrow(() -> new ResourceNotFoundException("Restaurante no encontrado"));
        }
        if (request.getProductoId() != null) {
            producto = productoRepository.findById(request.getProductoId())
                    .orElseThrow(() -> new ResourceNotFoundException("Producto no encontrado"));
        }
        if (restaurante == null && producto == null) {
            throw new BadRequestException("La alerta debe estar asociada a un restaurante o a un producto");
        }

        Alerta alerta = Alerta.builder()
                .usuario(usuario)
                .tipo(request.getTipo())
                .restaurante(restaurante)
                .producto(producto)
                .descripcion(request.getDescripcion())
                .estado(EstadoAlerta.PENDIENTE)
                .build();

        return AlertaResponse.desde(alertaRepository.save(alerta));
    }

    public List<AlertaResponse> listarDeUsuario(Long usuarioId) {
        return alertaRepository.findByUsuarioIdOrderByFechaDesc(usuarioId)
                .stream().map(AlertaResponse::desde).toList();
    }

    public List<AlertaResponse> listarPorEstado(EstadoAlerta estado) {
        List<Alerta> alertas = estado != null
                ? alertaRepository.findByEstadoOrderByFechaDesc(estado)
                : alertaRepository.findAll();
        return alertas.stream().map(AlertaResponse::desde).toList();
    }

    @Transactional
    public AlertaResponse revisar(Long id, RevisarAlertaRequest request, Usuario admin) {
        Alerta alerta = alertaRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Alerta no encontrada"));

        if (request.getEstado() == EstadoAlerta.PENDIENTE) {
            throw new BadRequestException("El estado de revision debe ser ACEPTADA o DENEGADA");
        }

        alerta.setEstado(request.getEstado());
        alerta.setRevisadoPor(admin);
        alerta.setFechaRevision(LocalDateTime.now());

        return AlertaResponse.desde(alertaRepository.save(alerta));
    }
}

File: frontend/lib/core/services/alerta_service.dart
import '../api/api_client.dart';
import '../models/alerta.dart';

class AlertaService {
  final ApiClient _client;
  AlertaService(this._client);

  Future<Alerta> crear({required TipoAlerta tipo, int? restauranteId, int? productoId, required String descripcion}) async {
    final data = await _client.post('/alertas', body: {
      'tipo': tipo.name,
      'restauranteId': restauranteId,
      'productoId': productoId,
      'descripcion': descripcion,
    });
    return Alerta.fromJson(data);
  }

  Future<List<Alerta>> misAlertas() async {
    final data = await _client.get('/alertas/me');
    return (data as List).map((e) => Alerta.fromJson(e)).toList();
  }

  // ---- Administracion ----

  Future<List<Alerta>> listarPorEstado({EstadoAlerta? estado}) async {
    final data = await _client.get('/admin/alertas', query: {'estado': estado?.name});
    return (data as List).map((e) => Alerta.fromJson(e)).toList();
  }

  Future<Alerta> revisar(int id, EstadoAlerta estado) async {
    final data = await _client.put('/admin/alertas/$id/revisar', body: {'estado': estado.name});
    return Alerta.fromJson(data);
  }
}
